pkg/karmadactl/addons/metricsadapter/metricsadapter.go

```diff
@@ -0,0 +1,206 @@
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	apiregistrationv1helper "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1/helper"
+
+	addoninit "github.com/karmada-io/karmada/pkg/karmadactl/addons/init"
+	addonutils "github.com/karmada-io/karmada/pkg/karmadactl/addons/utils"
+	initkarmada "github.com/karmada-io/karmada/pkg/karmadactl/cmdinit/karmada"
+	cmdutil "github.com/karmada-io/karmada/pkg/karmadactl/util"
+)
+
+// aaAPIServiceName define apiservice name install on karmada control plane
+var aaAPIServices = []string{
+	"v1beta1.metrics.k8s.io",
+	"v1beta1.custom.metrics.k8s.io",
+	"v1beta2.custom.metrics.k8s.io",
+}
+
+// AddonMetricsAdapter describe the metrics-adapter addon command process
+var AddonMetricsAdapter = &addoninit.Addon{
+	Name:    addoninit.MetricsAdapterResourceName,
+	Status:  status,
+	Enable:  enableMetricsAdapter,
+	Disable: disableMetricsAdapter,
+}
+
+var status = func(opts *addoninit.CommandAddonsListOption) (string, error) {
+	// check karmada-metrics-adapter deployment status on host cluster
+	deployClient := opts.KubeClientSet.AppsV1().Deployments(opts.Namespace)
+	deployment, err := deployClient.Get(context.TODO(), addoninit.MetricsAdapterResourceName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return addoninit.AddonDisabledStatus, nil
+		}
+		return addoninit.AddonUnknownStatus, err
+	}
+	if deployment.Status.Replicas != deployment.Status.ReadyReplicas ||
+		deployment.Status.Replicas != deployment.Status.AvailableReplicas {
+		return addoninit.AddonUnhealthyStatus, nil
+	}
+
+	// check metrics.k8s.io apiservice is available on karmada control plane
+	for _, aaAPIServiceName := range aaAPIServices {
+		apiService, err := opts.KarmadaAggregatorClientSet.ApiregistrationV1().APIServices().Get(context.TODO(), aaAPIServiceName, metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return addoninit.AddonDisabledStatus, nil
+			}
+			return addoninit.AddonUnknownStatus, err
+		}
+		if !apiregistrationv1helper.IsAPIServiceConditionTrue(apiService, apiregistrationv1.Available) {
+			return addoninit.AddonUnhealthyStatus, nil
+		}
+	}
+
+	return addoninit.AddonEnabledStatus, nil
+}
+
+var enableMetricsAdapter = func(opts *addoninit.CommandAddonsEnableOption) error {
+	if err := installComponentsOnHostCluster(opts); err != nil {
+		return err
+	}
+
+	if err := installComponentsOnKarmadaControlPlane(opts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var disableMetricsAdapter = func(opts *addoninit.CommandAddonsDisableOption) error {
+	// delete karmada metrics adapter service on host cluster
+	serviceClient := opts.KubeClientSet.CoreV1().Services(opts.Namespace)
+	if err := serviceClient.Delete(context.TODO(), addoninit.MetricsAdapterResourceName, metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	klog.Infof("Uninstall karmada metrics adapter  service on host cluster successfully")
```

Comment:
nits: extra blank space in `adapter   service`

Reply:
Updated.